resources/lib/modules/contextmenu.py

```diff
@@ -0,0 +1,213 @@
+
+        if not addons:
+            # Channel was not found.
+            _LOGGER.debug('No Add-on was found to play %s', program.get('channel'))
+            kodiutils.notification(message=kodiutils.localize(30710, channel=program.get('channel')))  # Could not find an Add-on...
+            return
+
+        if len(addons) == 1:
+            # Channel has one Add-on. Play it directly.
+            _LOGGER.debug('One Add-on was found to play %s: %s', program.get('channel'), addons)
+            cls._play(addons.values()[0], program)
+            return
+
+        # Ask the user to pick an Add-on
+        _LOGGER.debug('Multiple Add-on were found to play %s: %s', program.get('channel'), addons)
+        addons_list = list(addons)
+        ret = kodiutils.select(heading=kodiutils.localize(30711), options=addons_list)  # Select an Add-on...
+        if ret == -1:
+            _LOGGER.debug('The selection to play an item from %s was canceled', program.get('channel'))
+            return
+
+        cls._play(addons.get(addons_list[ret]), program)
+
+    @classmethod
+    def _play(cls, uri, program):
+        """Play the selected program with the specified URI."""
+        if '{date}' in uri:
+            uri = uri.format(date=program.get('start').isoformat())
+
+        if '{duration}' in uri:
+            uri = uri.format(date=program.get('duration').isoformat())
+
+        _LOGGER.debug('Playing "%s"', uri)
+        kodiutils.execute_builtin('PlayMedia', uri)
+
+    @classmethod
+    def get_selection(cls):
+        """Retrieve information about the selected ListItem."""
+
+        # The selected ListItem should be in sys.listitem, but I could not find real data in there that we can use for EPG selections.
+        # Therefore, we use xbmc.getInfoLabel(ListItem.xxx), but that references the currently selected ListItem. This could not be the same as the
+        # item where the Context Menu was opened on, when the selection was moved really quick before the Python code was started.
+        # It seems I'm not alone in this: https://forum.kodi.tv/showthread.php?tid=294357
+
+        # cls._debug_get_selection()
+
+        # Check if the selected item is also the intended item
+        if sys.listitem.getPath() != kodiutils.get_info_label('ListItem.FolderPath'):  # pylint: disable=no-member
+            # We are in trouble. We know that the data we want to use is invalid, but there is nothing we can do.
+            kodiutils.ok_dialog(message=kodiutils.localize(30712))  # Could not determine the selected program...
+            return None
+
+        # Load information from the ListItem
+        date = kodiutils.get_info_label('ListItem.Date')
+        duration = kodiutils.get_info_label('ListItem.Duration')
+        channel = kodiutils.to_unicode(kodiutils.get_info_label('ListItem.ChannelName'))
+
+        # Parse begin to a datetime
+        try:
+            start = datetime.strptime(date, '%d-%m-%Y %H:%M')
```

Comment:
`ListItem.Date` can contain all kinds of date formats. You should use `xbmc.getRegion` to determine the dateformat first.